File: 213-house-robber-ii/house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        nums1 = nums[1:]
        nums2 = nums[:-1]
        print(nums1, nums2)
        dic1 = {}
        dic2 = {}
        def helper(i,numz,dic):
            if i >= len(numz):
                return 0

            if i in dic:
                return dic[i]
            m = 0
            for r in range(i+2, len(numz)):
                #right neighbor
                n = helper(r,numz, dic)
                if n > m:
                    m = n
            dic[i] = m + numz[i]

            return dic[i]

        #Edge cases: Length of arr == 1
        #            Length of arr == 2
        #            Length of arr == 3

        if len(nums) == 1:
            return nums[0]
        elif len(nums) == 2:
            return max(nums[0], nums[1])
        
        return max(max(helper(0,nums1,dic1), helper(1,nums1,dic1)),max(helper(0,nums2,dic2), helper(1,nums2,dic2)))
        


        

File: 213-house-robber-ii/house-robber-ii.cpp
class Solution {
public:
    // // TC -> O(2 ^ N)
    // // SC -> O(1)
    // int recursive(vector<int>& nums, int index){
    //     if (index < 0){
    //         return 0;
    //     }
    //     if (index == 0){
    //         return nums[0];
    //     }

    //     int not_take = recursive(nums, index - 1);
    //     int take = nums[index] + recursive(nums,index - 2);

    //     return max(not_take, take);
    // }

    // // TC -> O(N)
    // // SC -> O(N + N)
    // int memoized(vector<int>& nums,vector<int>& dp, int index){
    //     if (index < 0) return 0;
    //     if (index == 0) return nums[0];
    //     if (dp[index] != -1) return dp[index];

    //     int not_take = memoized(nums, dp, index - 1);

    //     int take = nums[index] +  memoized(nums, dp, index - 2);

    //     dp[index] = max(not_take, take);

    //     return dp[index];
    // }

    int tabulation(vector<int>& nums){
        vector<int> dp(nums.size(), -1);
        dp[0] = nums[0];

        for(int i = 1; i < nums.size(); i++){

            int not_take = dp[i - 1];
            int take = nums[i];
            if (i - 2 >= 0){
                take = nums[i] + dp[i - 2];
            }

            dp[i] = max(take, not_take);
        }

        return dp[nums.size() - 1];
    }
    int rob(vector<int>& nums) {
        if (nums.size() == 1){
            return nums[0];
        }
        vector<int> omit_f(nums.begin()+1, nums.end());
        vector<int> omit_l(nums.begin(), nums.begin() + nums.size() - 1);
        return max(tabulation(omit_f), tabulation(omit_l));
    }
};